File: scala/src/main/scala/com/ngt/streamingwithflink/_01_AverageSensorReadings.scala
package com.ngt.streamingwithflink


import com.ngt.streamingwithflink.util.{SensorReading, SensorSource}
import org.apache.flink.api.common.eventtime.{BoundedOutOfOrdernessWatermarks, SerializableTimestampAssigner, WatermarkStrategy}
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.scala.function.WindowFunction
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector

import java.time.Duration

/**
 * @author ngt
 * @create 2021-05-07 21:45
 */
object AverageSensorReadings {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    // In Flink 1.12 the default stream time characteristic has been changed to TimeCharacteristic.EventTime,
    // 每一秒生成一次水位线
    // env.getConfig.setAutoWatermarkInterval(1000L)

    val sensorData: DataStream[SensorReading] = env
      .addSource(new SensorSource)
      .assignTimestampsAndWatermarks(WatermarkStrategy
        .forBoundedOutOfOrderness[SensorReading](Duration.ZERO)
        .withTimestampAssigner(new SerializableTimestampAssigner[SensorReading] {
          override def extractTimestamp(element: SensorReading, recordTimestamp: Long): Long = element.timestamp
        }))

    // 输入元素的时间戳如果是单调递增可以使用该简便方法
    //     .assignAscendingTimestamps(_.timestamp)


    // 输入元素的时间戳如果是单调递增也可以使用该方法
    //      .assignTimestampsAndWatermarks(WatermarkStrategy.forMonotonousTimestamps()
    //        .withTimestampAssigner(new SerializableTimestampAssigner[SensorReading] {
    //          override def extractTimestamp(t: SensorReading, l: Long): Long = t.timestamp
    //        }))

    // 输入元素的时间戳存在乱序，使用该方法
    //    .assignTimestampsAndWatermarks(WatermarkStrategy
    //      .forBoundedOutOfOrderness[SensorReading](Duration.ofSeconds(1))
    //      .withTimestampAssigner(new SerializableTimestampAssigner[SensorReading] {
    //        override def extractTimestamp(element: SensorReading, recordTimestamp: Long): Long = element.timestamp
    //      }))

    //.assignTimestampsAndWatermarks(new SensorTimeAssigner)  该方法已经过时


    val avgTemp: DataStream[SensorReading] = sensorData
      .map(r =>
        SensorReading(r.id, r.timestamp, (r.temperature - 32) * (5.0 / 9.0)))
      .keyBy(_.id)
      .window(TumblingEventTimeWindows.of(Time.seconds(1)))
      .apply(new TemperatureAverager)

    avgTemp.print()

    env.execute("Compute average sensor temperature")
  }
}

class TemperatureAverager extends WindowFunction[SensorReading, SensorReading, String, TimeWindow] {

  override def apply(
                      sensorId: String,
                      window: TimeWindow,
                      vals: Iterable[SensorReading],
                      out: Collector[SensorReading]): Unit = {

    val (cnt, sum) = vals.foldLeft((0, 0.0))((c, r) => (c._1 + 1, c._2 + r.temperature))
    val avgTemp = sum / cnt

    out.collect(SensorReading(sensorId, window.getEnd, avgTemp))
  }
}

File: java/src/main/java/com/ngt/streamingwithflink/_01_AverageSensorReadings.java
package com.ngt.streamingwithflink;

import com.ngt.streamingwithflink.util.SensorReading;
import com.ngt.streamingwithflink.util.SensorSource;
import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.windowing.WindowFunction;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;

import javax.xml.crypto.Data;
import java.time.Duration;

/**
 * @author ngt
 * @create 2021-05-17 21:28
 */
public class _01_AverageSensorReadings {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        SingleOutputStreamOperator<SensorReading> sensorData = env
                .addSource(new SensorSource())
                .assignTimestampsAndWatermarks(WatermarkStrategy
                        .<SensorReading>forBoundedOutOfOrderness(Duration.ZERO)
                        .withTimestampAssigner(new SerializableTimestampAssigner<SensorReading>() {
                            @Override
                            public long extractTimestamp(SensorReading element, long recordTimestamp) {
                                return element.timestamp;
                            }
                        }));


        env.addSource(new SensorSource())
                .assignTimestampsAndWatermarks(WatermarkStrategy.
                        <SensorReading>forMonotonousTimestamps()
                        .withTimestampAssigner((data, timestamp) -> data.timestamp));


        sensorData
                .map(value -> new SensorReading(value.id, value.timestamp, (value.temperature - 32) * (5.0 / 9.0)))
                .keyBy(data -> data.id)
                .window(TumblingEventTimeWindows.of(Time.seconds(1)))
                .apply(new TemperatureAverager())
                .print();

        env.execute("Compute average sensor temperature");

    }


    private static class TemperatureAverager implements WindowFunction<SensorReading, SensorReading, String, TimeWindow> {
        @Override
        public void apply(String s,
                          TimeWindow window,
                          Iterable<SensorReading> input,
                          Collector<SensorReading> out) throws Exception {

            int cnt = 0;
            double sum = 0.0;
            for (SensorReading r : input) {
                cnt++;
                sum += r.temperature;
            }
            double avgTemp = sum / cnt;

            out.collect(new SensorReading(s, window.getEnd(), avgTemp));
        }
    }

}

